gh-100414: Add SQLite backend to dbm

erlend-aasland opened the pull request:
Initial version authored by Raymond in https://github.com/python/cpython/issues/100414#issuecomment-1368273140.

<!-- gh-issue-number: gh-100414 -->
* Issue: gh-100414
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--114481.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

erlend-aasland committed:
gh-100414: Add SQLite backend to dbm

erlend-aasland committed:
Preliminary docs

erlend-aasland committed:
Decode path a little bit earlier

erlend-aasland commented on Doc/library/dbm.rst:
```diff
@@ -129,6 +134,69 @@ then prints out the contents of the database::
+   Open existing database for reading and writing
+
+.. |flag_c| replace::
+   Open database for reading and writing, creating it if it doesn't exist.
```
All of this is duplicated for each dbm submodule. Instead of copying the text yet another time, set up replacements and use these instead. In a follow-up docs PR, we can reuse these for the other modules.

erlend-aasland replied:
Added in 8c265408c51609c6b4a6788cac9cc5fea7a14888.

erlend-aasland committed:
Merge branch 'main' into sqlite/dbm

erlend-aasland commented on Doc/library/dbm.rst:
```diff
@@ -136,6 +141,69 @@ then prints out the contents of the database::
+
+.. |flag_n| replace::
+   Always create a new, empty database, open for reading and writing.
+
```
```suggestion
```

erlend-aasland committed:
Substitutions are already in place

erlend-aasland commented on Doc/library/dbm.rst:
```diff
@@ -136,6 +141,55 @@ then prints out the contents of the database::
+         * - ``'c'``
+           - |flag_c|
+         * - ``'n'``
+           - |flag_n|
```
```suggestion
      .. csv-table::
         :header: "Value", "Meaning"

         ``'r'`` (default), |flag_r|
         ``'w'``, |flag_w|
         ``'c'``, |flag_c|
         ``'n'``, |flag_n|
```

erlend-aasland committed:
Use CSV-style table, as the rest of the .rst file does

erlend-aasland committed:
dbm.sqlite3.error is a subclass of OSError

erlend-aasland committed:
Don't raise DB API exceptions; test this

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Don't panic on double close(); add more tests

erlend-aasland committed:
Corruption tests: check all basic operations for flags 'r', 'w', 'c'

erlend-aasland committed:
Add read-only specific test

gpshead commented on Lib/dbm/__init__.py:
```diff
@@ -49,6 +49,13 @@ class error(Exception):
     ndbm = None
 
+try:
+    import sqlite3
```
People using `dbm` are only likely to use a single backend. It is better to delay the import of the specific implementation until it is going to be used.  I'd just always add dbm.sqlite3 to _names above and adjust the loop checking for ImportError within `open()` below if neecessary at all.  An ImportError coming from the attempt to load `dbm.sqlite3` when someone needs or asks for it is sufficient.

erlend-aasland replied:
Hm, yeah. That's also similar to how the other backends are handled; for example, `dbm.gnu` is always present in `_names`, even though not all systems have the `_gdbm` module.

erlend-aasland replied:
This reminds me we'd better have some more tests for `.whichdb()`, to make sure SQLite databases are autodetected properly.

erlend-aasland replied:
Addressed in b655473dbd199006945c2aab8e754684ba7f1fec and 95eba9348c8939cc714881b2ac08648aaf0bfad6.

erlend-aasland committed:
Address review: always import sqlite3; let ImportError deal with missing deps

erlend-aasland committed:
Test whichdb

erlend-aasland committed:
test namespacing

erlend-aasland committed:
Test that 'c' works even if you already have a database; make sure BUILD_TABLE never can raise a DB API exception

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Remove unneeded comma

erlend-aasland committed:
Close cursors explicitly for each query

corona10 commented on Lib/dbm/sqlite3.py:
```diff
@@ -0,0 +1,117 @@
+  )
+"""
+GET_SIZE = "SELECT COUNT (key) FROM Dict"
+LOOKUP_KEY = "SELECT value FROM Dict WHERE key = ? LIMIT 1"
```
Hmm, since the key is the primary key, IMO, it should guarantee to return just one without limit.

erlend-aasland replied:
True that

erlend-aasland replied:
```suggestion
LOOKUP_KEY = "SELECT value FROM Dict WHERE key = ?"
```

corona10 reviewed (approved):
@erlend-aasland 
For me LGTM.
nit: Can we add the test for verification of unique key?
Something like..
```python
db["key"] = "foo"
db["key"] = "bar"
self.assertEqual(len(db), 1)
self.assertEqual(db["key"], "bar")
```

erlend-aasland commented on Lib/dbm/sqlite3.py:
```diff
@@ -0,0 +1,117 @@
+  CREATE TABLE IF NOT EXISTS Dict (
+    key TEXT NOT NULL,
+    value BLOB NOT NULL,
+    PRIMARY KEY (key)
```
IMO, we should let SQLite manage the primary key through the implicit `rowid`.

```suggestion
    key TEXT UNIQUE NOT NULL,
    value BLOB NOT NULL
```

erlend-aasland replied:
cc. @corona10

erlend-aasland replied:
- https://www.sqlite.org/lang_createtable.html

erlend-aasland replied:
I'll add some tests for key uniqueness and validity of the generated database.

erlend-aasland replied:
Ah, there is already generic tests for key uniqueness in `test_dbm.py`; we're covered.

serhiy-storchaka commented on Lib/dbm/__init__.py:
```diff
@@ -38,7 +38,7 @@
     pass
 
-_names = ['dbm.gnu', 'dbm.ndbm', 'dbm.dumb']
+_names = ['dbm.sqlite3', 'dbm.gnu', 'dbm.ndbm', 'dbm.dumb']
```
It makes dbm.sqlite3 the most preferable backend. Is it always better than dbm.gnu and dbm.ndbm?

erlend-aasland replied:
Yes, this is also one of Raymond's selling points in the OP (as I read him). I definitely think it should be the preferred backend.

serhiy-storchaka replied:
I think that Raymond said that it will be better than dbm.dumb, not other specialized engines.

presidento replied:
IMHO Sqlite is cross-platform (and Python has a many users with Windows), although likely not as fast as dbm.gnu or dbm.ndbm, since they are specialized engines especially for this use case.

So what does "better" mean? If better compatibility then Sqlite should be the first. If better performance, then it can be just before dbm.dumb.

For me both are acceptable, these are just my 50 cents.

erlend-aasland replied:
> I think that Raymond said that it will be better than dbm.dumb, not other specialized engines.

This is the wording from the OP; I let it speak for itself:

> Right now we support ndbm and gnu.dbm which might or might not be part of a given build. The fallback is the super slow dumbdbm. Not the sqlite3 is part of the standard build, we can do better.

Ping Raymond if you are unsure.

corona10 replied:
Hmm basicially dbm.gnu and dbm.ndbm is key/value storage product itself but sqlite3 is more rich product than them, I am also not sure that making sqlite3 to become a most preferrable engine is right choice at this moment.

erlend-aasland replied:
In terms of user experience, dbm.sqlite3 is the clear winner IMO. It is available for everyone; a user can create a database using `dbm.open()` on a Linux machine, copy the generated database to a Windows or macOS machine, open the database using `dbm.open()`, and it just works. Let's ask on Discourse.

corona10 replied:
IMO, even if we will migrate the most preferable engine to sqlite3, CPython 3.13 should be in the experimental stage. So, the most preferable engine should not be SQLite3 at this moment if we consider the experimental stage.
People can argue about different performance issues or behavior issues that can not be expected at this moment.
After CPython 3.13, we can make the decision with more context.

erlend-aasland replied:
cc. @rhettinger

serhiy-storchaka replied:
If "better" means better compatibility then dbm.dumb should be the first. I do not think Raymond used it in this meaning.

erlend-aasland replied:
In order to unlock this PR, let us go for this:

```suggestion
_names = ['dbm.gnu', 'dbm.ndbm', 'dbm.sqlite3', 'dbm.dumb']
```

I'll create a discussion on Discourse, and we can chose to promote it in a follow-up PR.

erlend-aasland replied:
I created a poll: https://discuss.python.org/t/new-default-preferred-dbm-backend/44228

serhiy-storchaka commented on Lib/dbm/sqlite3.py:
```diff
@@ -0,0 +1,117 @@
+        # We use the URI format when opening the database.
+        path = os.fsdecode(path)
+        if not path or path == ":memory:":
+            uri = "file:?mode=memory"
```
It should be documented.

erlend-aasland replied:
Yes, it should; thanks!

erlend-aasland replied:
Perhaps we should not allow it; this is not sqlite3, it is dbm. I suggest we remove this "feature" and treat everything as an ordinary filename.

erlend-aasland replied:
There is enough time till the next alpha to decide if we want this or not; let's leave it out for now, at least.

serhiy-storchaka replied:
Also, does it work for filenames like "localhost" or "/localhost"?

erlend-aasland replied:
> Also, does it work for filenames like "localhost" or "/localhost"?

Those are not special names; they will be accepted as ordinary file names.

erlend-aasland replied:
Let me take that back; `"localhost"` _is_ considered special by SQLite. Let's add more tests.

serhiy-storchaka commented on Lib/dbm/sqlite3.py:
```diff
@@ -0,0 +1,117 @@
+    pass
+
+
+def _normalize_uri_path(path):
```
What if filename contains "%"?

erlend-aasland replied:
Let's find out; I implemented ~~all~~ most of SQLite's recommendations for URI normalisation: https://sqlite.org/uri.html

I see I did not address the backslash case; I'll fix that and add more tests.

erlend-aasland replied:
I rewrote this using pathlib and also added substitution for "%".

erlend-aasland replied:
Windows paths are ... irritating. ~~I'll have to look at those later; for now, I marked the Windows tests as WIP.~~ Windows tests are enabled; crossing fingers 🤞

erlend-aasland replied:
Pushed 08c38481c10d8cbfcc9ee02f2133a3285eed216b which just uses pathlib's `as_uri()` with some amendment. I guess it's going to break on Windows. Anyway, I'm calling it a day. Thanks for the input so far!

serhiy-storchaka commented on Lib/dbm/sqlite3.py:
```diff
@@ -0,0 +1,117 @@
+
+    def __delitem__(self, key):
+        if key not in self:
+            raise KeyError(key)
```
It is not efficient and is not atomic. Can we determine whether deletion actually deleted any key by the query result?

erlend-aasland replied:
Yeah, I think we can implement this better; I'll address it. Thanks.

erlend-aasland replied:
> Can we determine whether deletion actually deleted any key by the query result?

We cannot use the RETURNING clause, as that is only available for SQLite 3.35.0 and newer. We _could_ verify it by querying again, but I don't think it is worth it; if DELETE FROM fails, we'll get an operational error, which in turn will show up as an exception for the user.

erlend-aasland replied:
Implemented by checking `Cursor.rowcount` after the attempted delete operation. Works swell.

serhiy-storchaka commented on Lib/dbm/sqlite3.py:
```diff
@@ -0,0 +1,117 @@
+
+    def __iter__(self):
+        with closing(self._execute(ITER_KEYS)) as cu:
+            for row in cu:
```
What happen when we close `_Database` in the middle of iteration? What happen when we close it before the first iteration? Are there tests for this?
```py
with closing(iter(db)) as it:
    db.close()
    next(it)  # ???
```
```py
with closing(iter(db)) as it:
    next(it)
    db.close()
    next(it)  # ???
```

erlend-aasland replied:
Let's add those tests.

erlend-aasland replied:
Note that the other implementations do not implement `__iter__`. We do, since it is required by the ABC we're using. Anyway, I added guards for misuse.

serhiy-storchaka commented on Lib/dbm/sqlite3.py:
```diff
@@ -0,0 +1,117 @@
+                    pass
+            case _:
+                raise ValueError("Flag must be one of 'r', 'w', 'c', or 'n', not",
+                                 repr(flag))
```
Should there be a formatting operation?

erlend-aasland replied:
Yes, of course, this is not `print()` :)

erlend-aasland committed:
Address review from Donghee, Serhiy, and myself:

- Amend dbm docstring
- Add misuse guards
- Let SQLite manage the rowid
- Atomic delete
- Hide connection member
- Remove :memory: feature
- Improve URI normalisation
- Fix ValueError formatting

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Add URI tests, fix whitespace, catch errors during db creation

erlend-aasland committed:
Enable test on Windows

erlend-aasland committed:
Use as_uri() instead

erlend-aasland committed:
Skip tests with relative paths for now

erlend-aasland committed:
Add prelim docstring for dbm.sqlite3.open

erlend-aasland committed:
Amend Windows URI expected results

felixxm commented on Lib/dbm/__init__.py:
```diff
@@ -164,6 +164,10 @@ def whichdb(filename):
         return ""
 
+    # Check for SQLite3 header string.
+    if s16 == b"SQLite format 3\0":
```
I know the current comparison works, but can we use `3\000` to be consistent with [SQLite docs](https://www.sqlite.org/fileformat.html#the_database_header):
```suggestion
    if s16 == b"SQLite format 3\000":
```

erlend-aasland replied:
Hm, I would rather do this instead; it feels more "pythonic":

```suggestion
    if s16 == b"SQLite format 3\x00":
```

But it is not a big issue for me; I'm fine either way.

felixxm commented on Lib/test/test_dbm.py:
```diff
@@ -213,6 +220,27 @@ def test_whichdb_ndbm(self):
 
+    @unittest.skipUnless(dbm_sqlite3, reason='Test requires dbm.sqlite3')
+    def test_whichdb_sqlite3(self):
+        # Test that databases created by dbm.sqlite3 are detected correctly.
```
I'm not sure it's worth adding prefixes like `Test that`  since all tests test things :upside_down_face: 
```suggestion
        # Databases created by dbm.sqlite3 are detected correctly.
```

felixxm reviewed (commented):
@erlend-aasland Thanks :+1: I left only two minor suggestions.

erlend-aasland committed:
For now, just remove the relative URI tests on Windows

erlend-aasland committed:
Document dbm.sqlite3.open signature

erlend-aasland committed:
Update Lib/test/test_dbm.py

Co-authored-by: Mariusz Felisiak <felisiak.mariusz@gmail.com>

felixxm commented on Lib/test/test_dbm.py:
```diff
@@ -213,6 +220,27 @@ def test_whichdb_ndbm(self):
+
+    @unittest.skipUnless(dbm_sqlite3, reason='Test requires dbm.sqlite3')
+    def test_whichdb_sqlite3_existing_db(self):
+        # Test that existing sqlite3 databases are detected correctly.
```
```suggestion
        # Existing sqlite3 databases are detected correctly.
```

erlend-aasland committed:
Update Lib/test/test_dbm.py

Co-authored-by: Mariusz Felisiak <felisiak.mariusz@gmail.com>

erlend-aasland committed:
Only test URI substitution and normalisation; discard prefix

erlend-aasland committed:
Pull in changes from PR

erlend-aasland committed:
Doc amendments

erlend-aasland committed:
Pull in main

serhiy-storchaka commented on Lib/dbm/__init__.py:
```diff
@@ -5,7 +5,7 @@
         d = dbm.open(file, 'w', 0o666)
 
-The returned object is a dbm.gnu, dbm.ndbm or dbm.dumb object, dependent on the
+The returned object is a dbm.sqlite3, dbm.gnu, dbm.ndbm or dbm.dumb object, dependent on the
```
```suggestion
The returned object is a dbm.sqlite3, dbm.gnu, dbm.ndbm or dbm.dumb database object, dependent on the
```

serhiy-storchaka commented on Lib/dbm/sqlite3.py:
```diff
@@ -0,0 +1,145 @@
+    path = os.fsdecode(path)
+    path = Path(path)
+    uri = path.absolute().as_uri()
+    while "//" in uri:
+        uri = uri.replace("//", "/")
```
Is it needed?

erlend-aasland replied:
Yes.

serhiy-storchaka commented on Lib/dbm/sqlite3.py:
```diff
@@ -0,0 +1,145 @@
+        if not self._cx:
+            raise error(_ERR_CLOSED)
+        try:
+            ret = self._cx.execute(*args, **kwargs)
```
It can be simply:
```suggestion
            return self._cx.execute(*args, **kwargs)
```
It is even better to wrap it in `closing()`:
```suggestion
            return closing(self._cx.execute(*args, **kwargs))
```

erlend-aasland replied:
Wrapping in closing is a nice trick; I will apply that.

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Mark up flags as list; it renders better

erlend-aasland committed:
Apply suggestions from code review

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

erlend-aasland committed:
Address review: use closing() in _execute() wrapper

erlend-aasland commented on Doc/library/dbm.rst:
```diff
@@ -141,6 +146,47 @@ then prints out the contents of the database::
+      The path to the database to be opened.
+   :type filename: :term:`path-like object`
+
+   :param str flag:
```
List markup needs a newline

```suggestion
   :param str flag:

```

erlend-aasland committed:
Update Doc/library/dbm.rst

erlend-aasland committed:
Add support for 'mode' param in dbm.sqlite3.open()

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Fix test_misuse_reinit()

erlend-aasland committed:
Merge branch 'main' into sqlite/dbm

erlend-aasland committed:
Address Serhiy's offline remark: coerce keys/values to bytes

erlend-aasland committed:
Align docs to e782fad38f

serhiy-storchaka commented on Lib/test/test_dbm_sqlite3.py:
```diff
@@ -208,7 +208,12 @@ def test_misuse_empty_filename(self):
+        # (raw, coerced)
+        (42, b"42"),
+        (3.14, b"3.14"),
+        ("string", b"string"),
```
Does it work with non-ASCII strings?

erlend-aasland replied:
Do you mean such a test?

```diff
diff --git a/Lib/test/test_dbm_sqlite3.py b/Lib/test/test_dbm_sqlite3.py
index 7bc2a03035..3d2dc5fd55 100644
--- a/Lib/test/test_dbm_sqlite3.py
+++ b/Lib/test/test_dbm_sqlite3.py
@@ -214,6 +214,7 @@ class DataTypes(_SQLiteDbmTests):
         (3.14, b"3.14"),
         ("string", b"string"),
         (b"bytes", b"bytes"),
+        ("æøå", b"\xc3\xa6\xc3\xb8\xc3\xa5"),
     )
 
     def setUp(self):
```

serhiy-storchaka reviewed (approved):
LGTM if all operations still accept strings.

erlend-aasland committed:
Compat with other backends: silently coerce keys to bytes
